### PR TITLE
refactor: change `enum Action` to `Arc<dyn ActionTrait>`, change file structure

### LINKED ISSUE #97: Refactor `enum Action` into `trait ActionTrait`
### Description

The goal of this issue is to restructure Actions and Reducers in such a manner that there is a clear format to adhere to. This will make it much clearer and easier for Impierce but also for third parties to add and navigate through the Actions and Reducers.

The idea is the have all Actions as structs with traits so that the Reducers will be coupled to the Actions as implementations.

The Reducers will remain in their separate file structure, as they currently are.

We will also use the `typetag` library to compliment these structs. This will make it possible to deserialize Actions from front-end payloads only by name of the struct. This will keep communication between front-end and back-end mostly as it was.

The files will be restructured as well.
One folder for actions and one file per action.
The mod file will contain the format (trait) specifications as well, supplemented with guidelines for third party usage.
The result will be a clearer structure of our own code but most certainly also a much clearer, cleaner and safer pattern to follow for third parties to add their code.

### Motivation

This will make sure we have a cleaner solution for both `Action`s as well as `Reducer`s that's closer to the Redux pattern. 

Implementing Actions with the help of the `typetag` crate will also allow for adding custom Actions 'from the outside', thus allows for extensibility.

### Resources

n/a

### To-do List

[ ] Define traits & macro rules in mod file
[ ] Transpose existing code to new format
[ ] Create file structure and add comments
[ ] Add guidelines to the mod file
[ ] Test new code

### PR DESCRIPTION
**Pull request by Norlock**
# Description of change
Directory structure has changed.

This change introduces the `trait ActionTrait`:
```rust
/// A trait that all Actions must implement.
#[typetag::serde(tag = "type", content = "payload")]
pub trait ActionTrait: Send + std::fmt::Debug + DowncastSync {
    /// Returns the reducers that should be called when this action is dispatched.
    fn reducers<'a>(&self) -> Vec<Reducer<'a>>;
}
```
This trait is implemented on all types that represent `Action`s. Currently our Actions are represented inside a `enum Action`, which means that the set of Actions is bounded (bounded by the size of the enum). The new `ActionTrait` trait (in combination with the `typetag` crate) allows for an unbounded amount of Actions.

This allows for the `handle_action` command handler to still be able to deserialize actions:
```rust
pub async fn handle_action<R: tauri::Runtime>(
    action: Action,                                     // <-- `Arc<dyn ActionTrait>`
    _app_handle: tauri::AppHandle<R>,
    container: tauri::State<'_, AppStateContainer>,
    window: tauri::Window<R>,
) -> Result<(), String>
```

The `ActionTrait` trait can be implemented like this:
```rust 
#[derive(serde::Serialize, serde::Deserialize, Debug, TS, Clone)]
#[ts(export, export_to = "bindings/actions/UnlockStorage.ts")]
pub struct UnlockStorage {
    pub password: String,
}

#[typetag::serde(name = "[Storage] Unlock")]
impl ActionTrait for UnlockStorage {
    fn reducers<'a>(&self) -> Vec<Reducer<'a>> {
        vec![reducer!(unlock_storage)]
    }
}
```

Where the `fn reducers` method returns a vector of function pointers pointing to specific reducer functions.

The action reducers 'listen' to incoming actions by trying to downcast the generic `Arc< dyn ActionTrait>` to an action implementation of that trait, for example:
```rust
pub async fn unlock_storage(state: AppState, action: Action) -> Result<AppState, AppError> {
    if let Some(password) = listen::<UnlockStorage>(action).map(|payload| payload.password) {
        ...
       return Ok(updated_state);
   }
   Ok(state)
}
```
If the incoming action cannot be downcasted, than the current state is returned without being updated.

So the other notable change to the reducer functions is that their prototype changed from:
```rust
pub async fn reducer_function(state: &mut AppState, action: Action) -> Result<(), AppError>
```

to:
```rust
pub async fn reducer_function(state: AppState, action: Action) -> Result<AppState, AppError>
```
This means that a reducer always takes the current state, and only returns an updated state when the entire reducer completes successfully. Otherwise, the same current unchanged state is returned. This solves the issue of the appstate not correctly being 'reset' first mentioned in #83.

## Links to any relevant issues
Closes #97

## How the change has been tested
Application been run on dev and in build, (desktop + android)
Existing tests still pass, successful manual testing on physical testing with the NGDIL demo.

## Definition of Done checklist
Add an `x` to the boxes that are relevant to your changes.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: unime/src-tauri/android-build/key.properties
Comment (daniel-mader) [inline unime/src-tauri/android-build/key.properties:1]: Please remove this and add it to `.gitignore`. I know it's fake values, but then add placeholders/variables instead.

File: unime/src-tauri/gen/android/app/proguard-rules.pro
Comment (daniel-mader) [inline unime/src-tauri/gen/android/app/proguard-rules.pro:1]: Why is this needed?

File: unime/src-tauri/android-build/build.gradle.kts [hunk 1]
```diff
@@ -27,16 +19,16 @@ android {
         applicationId = "com.impierce.identity_wallet"
         minSdk = 24
         targetSdk = 33
-        versionCode = 32
-        versionName = "0.4.7"
+        versionCode = 1
```
Comment (daniel-mader) [inline unime/src-tauri/android-build/build.gradle.kts]: Why are we resetting the `versionCode`?

File: unime/src/routes/prompt/password-required/+page.svelte [hunk 2]
```diff
@@ -78,7 +78,7 @@
           <div slot="content" class="w-full pb-[10px] pt-[20px]">
             <button
               class="h-[48px] w-full rounded-xl bg-rose-100 px-4 py-2 text-[14px]/[24px] font-medium text-rose-500"
-              on:click={() => dispatch({ type: '[App] Reset' })}>{$LL.SETTINGS.RESET_APP.CONFIRM()}</button
+              on:click={() => dispatch({ type: '[App] Reset', payload: null })}>{$LL.SETTINGS.RESET_APP.CONFIRM()}</button
```
Comment (daniel-mader) [inline unime/src/routes/prompt/password-required/+page.svelte]: Is a payload always required or can we default to null when no payload given?

File: unime/src/routes/prompt/accept-connection/+page.svelte [hunk 3]
```diff
@@ -39,7 +39,7 @@
   console.log($state.current_user_prompt);
 
   onDestroy(async () => {
-    dispatch({ type: '[User Flow] Cancel' });
+    dispatch({ type: '[User Flow] Cancel', payload: { } });
```
Comment (daniel-mader) [inline unime/src/routes/prompt/accept-connection/+page.svelte]: What's the difference between `payload: { }` and `payload: null`?

File: unime/src/routes/(app)/me/search/+page.svelte [hunk 4]
```diff
@@ -21,7 +21,14 @@
     <Search
       on:value={(e) => {
         searchTerm = e.detail;
-        dispatch({ type: '[User Data] Query', target: 'Credentials', search_term: e.detail });
+        dispatch({ 
+          type: '[User Data] Query',
+          payload: {
+            target: 'Credentials',
+            search_term: e.detail,
+            sort_reverse: false,
```
Comment (daniel-mader) [inline unime/src/routes/(app)/me/search/+page.svelte]: Can `sort_reverse` have a default of `false`, so it doesn't need to be explicitly passed?

File: unime/src-tauri/gen/android/app/proguard-rules.pro
Comment (nanderstabel) [reply unime/src-tauri/gen/android/app/proguard-rules.pro:1]: It turned out that the reason for the android build (`cargo tauri android build`) to crash was because of this being set to `true` in the `build.gradle.kts` file:
```
        getByName("release") {
            isMinifyEnabled = true
```

Minify optimizes and shrinks the code of the apk/bundle but with the new file structure, it optimized away the `RustWebView` causing the app to crash. Adding the last three lines to the generated `proguard-rules.pro` file prevents this.

File: unime/src-tauri/android-build/android-init.sh
Comment (daniel-mader) [inline unime/src-tauri/android-build/android-init.sh:1]: I wonder if we really need such a script in the repo. It might save running a few commands once, but you shouldn't need to run `cargo tauri android init` often on your local machine anyway. I would rather refer to the official Tauri docs here.

Comment (nanderstabel) [reply unime/src-tauri/android-build/android-init.sh:1]: :heavy_plus_sign:

File: unime/src-tauri/gen/apple/ExportOptions.plist [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+		<key>com.impierce.identity-wallet</key>
+		<string>UniMe (alpha) - Internal Test (Development)</string>
+	</dict> -->
+
+    <key>method</key>
+	<string>ad-hoc</string>
+	<key>provisioningProfiles</key>
+	<dict>
+		<key>com.impierce.identity-wallet</key>
+		<string>UniMe (alpha) - Internal Test (Distribution)</string>
```
Comment (nanderstabel) [inline unime/src-tauri/gen/apple/ExportOptions.plist]: @daniel-mader , can we actually change `<key>com.impierce.identity-wallet</key>` to `<key>com.impierce.unnime</key>`  do you know?
